libgnucash/engine/qof-string-cache.cpp

```diff
@@ -62,48 +62,38 @@ qof_string_cache_clear (void)
 void
 qof_string_cache_remove(const char * key)
 {
-    if (key && *key)
-    {
-        auto cache_iter = qof_string_cache.find (key);
+    if (!key || !key[0])
+        return;
 
-        if (cache_iter == qof_string_cache.end())
-            return;
+    auto cache_iter = qof_string_cache.find (key);
+    if (cache_iter == qof_string_cache.end())
+        return;
 
-        auto& refcount = cache_iter->second;
-
-        if (refcount == 1)
-            qof_string_cache.erase (cache_iter);
-        else
-            refcount--;
-    }
+    auto& refcount = cache_iter->second;
+    if (refcount == 1)
+        qof_string_cache.erase (cache_iter);
+    else
+        refcount--;
 }
 
 /* If the key exists in the cache, increment the refcount.  Otherwise,
  * add it with a refcount of 1. */
 const char *
 qof_string_cache_insert(const char * key)
 {
-    if (key)
-    {
-        if (key[0] == 0)
-        {
-            return "";
-        }
+    if (!key)
+        return nullptr;
 
-        auto cache_iter = qof_string_cache.find (key);
+    if (!key[0])
+        return "";
 
-        if (cache_iter == qof_string_cache.end())
-            cache_iter = qof_string_cache.insert_or_assign (key, 1).first;
-        else
-        {
-            auto& refcount = cache_iter->second;
-            refcount++;
-        }
+    auto cache_iter = qof_string_cache.find (key);
+    if (cache_iter == qof_string_cache.end())
+        cache_iter = qof_string_cache.insert_or_assign (key, 1).first;
+    else
+        (cache_iter->second)++;
 
```

Comment:
How about:
```
auto [cache_iter, result] = qof_string_cache.insert(key, 0);
(cache_iter->second)++;
```

`insert` will return an iter to the pre-existing map element for key if it existed already or an iter to the newly inserted element if it didn't exist yet. By initializing a new value to 0, incrementing the value by 1 does the right thing regardless of the iter that's returned. For new elements the net result is the value is 1, for pre-existing elements the value is increased by one.

I think this will save iterating over the map twice (once for find and once for insert_or_assign).

Note the `auto [ a, b]` notation is a new feature in C++17 and allows to assign elements from a pair to individual variables.

Reply:
If you use C++17 structured binding then `cache_iter` isn't a pair any more and `cache_iter->second` doesn't exist. It's just `++cache_iter`.

Reply:
> ```
> bool result;
> [cache_iter, result] = qof_string_cache.insert_or_assign({key, 1});
> ```
> 

It seems to me you're mixing up `insert` and `insert_or_assign` syntax. `insert` can use an initializer list to create a map element, `insert_or_assign` is a function with two parameters `key` and  `value`.
While not to the point it's confusing me here as I deliberately chose to use `insert` instead of `insert_or_assign` and it's not clear whether you understood why.

> is clearer than
> 
> ```
> cache_iter = qof_string_cache.insert_or_assign (key, 1).first;
> ```
> 
That I do agree to.

> Plus in either case you should test `result` to make sure that the insertion succeeded.

Why ? The way I [interpret the description in cppreference](https://en.cppreference.com/w/cpp/container/unordered_map/insert) this particular invocation of insert will always return an iterator that points to an element in the map with the key we attempted to insert. If insertion fails it will be the element that already existed with this key, otherwise it will be the element that got inserted. For our particular use case it doesn't matter which of these two it is - our next step is always to increment its counter.
Do you think my interpretation is wrong ?

Reply:
> For our particular use case it doesn't matter which of these two it is

OK.